pkg/fs/tail.go

```diff
@@ -0,0 +1,215 @@
+package fs
+
+import (
+	"errors"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+var ErrIsDir = errors.New("path is a directory")
+
+type TFile struct {
+	f       *os.File
+	watcher *fsnotify.Watcher
+}
+
+func TailFile(name string) (*TFile, error) {
+	info, err := os.Stat(name)
+	if err != nil {
+		return nil, err
+	}
+	if info.IsDir() {
+		return nil, ErrIsDir
+	}
+	f, err := OpenFile(name, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		f.Close()
+		return nil, err
+	}
+	if err := watcher.Add(name); err != nil {
+		f.Close()
+		watcher.Close()
+		return nil, err
+	}
+	return &TFile{f, watcher}, nil
+}
+
+func (t *TFile) Read(b []byte) (int, error) {
+read:
+	n, err := t.f.Read(b)
+	if errors.Is(err, os.ErrClosed) {
+		err = io.EOF
+	}
+	// If there is data and EOF, change err to nil so data is read. Assuming
+	// no data is added to the file, the next Read call will initiate the wait.
+	if n > 0 && err == io.EOF {
+		err = nil
+	}
+	if n == 0 && err == io.EOF {
+		if err := t.waitWrite(); err != nil {
+			return 0, err
+		}
+		goto read
+	}
```

Comment:
Nits: Wrap the io.EOF handling in a single if statement. And save a trip through t.waitWrite if t.f.Read returns 0 and os.ErrClosed by handling io.EOF before os.ErrClosed.
```suggestion
	if err == io.EOF {
		if n > 0 {
			return n, nil
		}
		if err := t.waitWrite(); err != nil {
			return 0, err
		}
		goto read
	}
	if errors.Is(err, os.ErrClosed) {
		err = io.EOF
	}
```

Reply:
genius!